### PR TITLE
Mark messages read on stream unsubscribe

### PR DESCRIPTION
**Pull request by hdorothea**

### CODE REVIEW

**Comment by timabbott** [comment]
@hdorothea thanks for working on this!  I think possibly the correct implementation here might be to make this change on the Zulip backend, however; since I think we have the same bug if you unsubscribe from the webapp (though note the symptom is invisible, because of how the webapp accesses data differently).  See https://github.com/zulip/zulip/issues/5959 for the issue we recently opened for this problem.

File: src/streams/StreamList.js [hunk 1]
```diff
@@ -77,8 +77,14 @@ export default class StreamList extends PureComponent {
             showSwitch={showSwitch}
             isSwitchedOn={item.subscribed}
             onPress={onPress}
-            onSwitch={onSwitch}
-          />}
+            onSwitch={
```
Comment (borisyankov) [inline src/streams/StreamList.js]: We do not inline function declarations inside a component like this.
The reason is that a new function is created for each component render. This not only wasted memory but also forces a repaint of the item, even when no other props have changed.

In this case, it also creates multiple functions per render as there is a list of StreamItems being rendered.

The solution is a `handleSwitch` method and then you do `onSwitch={this.handleSwitch}`

File: src/subscriptions/SubscriptionsCard.js [hunk 2]
```diff
@@ -23,13 +24,14 @@ export default class SubscriptionsCard extends PureComponent {
 
   handleFilterChange = (filter: string) => this.setState({ filter });
 
-  handleSwitchChange = (streamName: string, switchValue: boolean) => {
-    const { auth } = this.props;
+  handleSwitchChange = (streamName: string, switchValue: boolean, streamId: number) => {
+    const { auth, markStreamMessagesRead } = this.props;
 
     if (switchValue) {
       subscriptionAdd(auth, [{ name: streamName }]);
```
Comment (borisyankov) [inline src/subscriptions/SubscriptionsCard.js]: We would likely need to handle the adding of unread messages when you subscribe again?

File: src/subscriptions/SubscriptionsContainer.js [hunk 3]
```diff
@@ -5,12 +5,25 @@ import type { GlobalState } from '../types';
...
 import SubscriptionsCard from './SubscriptionsCard';
+import { getUnreadStreams } from '../baseSelectors';
+import { queueMarkAsRead } from '../api';
 
 export default connect(
   (state: GlobalState) => ({
     auth: getAuth(state),
     streams: state.streams,
     subscriptions: state.subscriptions,
+    markStreamMessagesRead: (streamId: number) => {
```
Comment (borisyankov) [inline src/subscriptions/SubscriptionsContainer.js]: This is a place where we specify what parts of the Redux state we want.
Creating a new function wouldn't go here.

File: src/subscriptions/SubscriptionsContainer.js [hunk 4]
```diff
@@ -5,12 +5,25 @@ import type { GlobalState } from '../types';
...
+      const removeIds = [].concat(
```
Comment (borisyankov) [inline src/subscriptions/SubscriptionsContainer.js]: You are making two unnecessary array copies here.
The first one is the `[].concat()` and the second is the `...` in front of `getUnreadStreams`.
'filter' already creates a new array so `getUnreadStreams(state).filter()` is enough.

File: src/subscriptions/SubscriptionsContainer.js [hunk 5]
```diff
@@ -5,12 +5,25 @@ import type { GlobalState } from '../types';
...
+        ...getUnreadStreams(state)
+          .filter(stream => streamId === stream.stream_id)
+          .map(stream => stream.unread_message_ids),
+      );
```
Comment (borisyankov) [inline src/subscriptions/SubscriptionsContainer.js]: Optionally, but we tend to do it now, we do a check of `length` of input vs output after the `filter`. If the lengths didn't change, we do not need to use the new value as it did not change.

File: src/subscriptions/SubscriptionsContainer.js [hunk 6]
```diff
@@ -5,12 +5,25 @@ import type { GlobalState } from '../types';
...
+
+      if (removeIds.length > 0) {
+        queueMarkAsRead(getAuth(state), removeIds);
```
Comment (borisyankov) [inline src/subscriptions/SubscriptionsContainer.js]: We would likely want to do the actual 'is not unread' logic in a different way. 
We don't want to mark this on the server, but just not display the messages as unread.

**Comment by hdorothea** [comment]
​
Thanks for the feedback, it was very helpful.

I agree the bigger issue is: what should happen to read/unread on unsubscribe. And that should be discussed in the main repository issue and implemented on the backend.

As to the question of wether we should have a simple mobile front-end fix: It would be trivial to remove the unsubscribed stream from unread.streams in the state and thus from the unread-tab. However adding it back there on resubscribe with consistent behavior on reload seems hard as, as far as I can get it the unread message ids are (only?) [sent with the initial state](https://github.com/zulip/zulip/blob/274938bf5fb0eef347924e53634d4472ebc83a72/docs/unread_messages.md).

File: src/subscriptions/SubscriptionsContainer.js (hunk 4 above)
Comment (hdorothea) [reply src/subscriptions/SubscriptionsContainer.js]: ​state.unread.streams is an array of topic objects, which is why I potentially get a nested array which I am flattening here.

**Comment by borisyankov** [comment]
@timabbott @showell is this resolved server-side now?

**Comment by showell** [comment]
Not yet.  We've made the decision to mark messages as read when you unsubscribe, but it hasn't been implemented yet.

**Comment by borisyankov** [comment]
I will close this PR then. We will auto-magically have this fixed in the mobile app when this lands in the backend.

@hdorothea thanks for noticing this and working on the issue. I would love it if you continue working on anything else you want to fix or improve upon.

**Pull request closed without merging**